### ISSUE TITLE
Failure to compile on MacOS 13 with Apple Clang

### ISSUE THREAD

**Issue by ktbarrett** [verilator/verilator#5190]
Failure [here](https://github.com/cocotb/cocotb/actions/runs/9598197748/job/26469167989?pr=3952#step:24:4132). We are installing verilator v5.026 using `brew install verilator`. It seems to be able to include the coroutine library, but things are *not* made available in the `std` namespace.

This worked fine in v5.024.

**Comment by wsnyder** [comment]
Likely related to #5030. Please read that, then ideally propose a pull as it won't be able to be fixed otherwise, thanks.

**Comment by simon-cogni** [comment]
I hit this issue also, in my setup of MacOS Sonoma 14.5, Apple clang 15.0.0, Homebrew install of Verilator (built from source, using default system toolchain).  

As I understand it, verilator's configure.ac is not testing -std=gnu++20 (yet) and therefore only sets "-std=gnu++17" to clang.  i.e. : 

# Flag to select newest language standard supported
# Macros work such that first option that passes is the one we take
# Currently enable c++17/c++14 due to packaged SystemC dependency
# c++17 is the newest that Verilator is regularly tested to support
# c++14 is the oldest that Verilator supports
# gnu is required for Cygwin to compile verilated.h successfully
#_MY_CXX_CHECK_SET(CFG_CXXFLAGS_STD_NEWEST,-std=gnu++20)
#_MY_CXX_CHECK_SET(CFG_CXXFLAGS_STD_NEWEST,-std=c++20)
_MY_CXX_CHECK_SET(CFG_CXXFLAGS_STD_NEWEST,-std=gnu++17)
_MY_CXX_CHECK_SET(CFG_CXXFLAGS_STD_NEWEST,-std=c++17)

Clang (at least apple) seems to default to c++98 standard when not given -std=xxx, so it makes sense that it always gets something in that flag.  

Meanwhile #include <coroutine> is a c++20 thing.   The __coroutine/coroutine_handle.h is wrapped in "#if _LIBCPP_STD_VER >= 20" in BOTH my MacOS and Arch Linux systems.  It's just that the Arch install uses g++, and is not given any -std=xxx flags, and is defaulting to c++17 semantics, as well as being given -fcoroutines, which enables the experimental coroutines support.  


If Verilator now uses coroutines, and the coroutine headers are requiring c++20 (at least in clang), then perhaps the configure.ac lines with the c++20 check should be uncommented (at least in clang)?   Otherwise it seems like clang support maybe broken (at least with --timing, which I believe is where coroutines are used).  

I don't have the knowledge to switch my Arch machine away from default verilator install options, not with enough confidence in the results that I'd need to share the results :) ...  but I did play with clang-17.0.6 on linux and confirmed that coroutine headers require -std=gnu++20.   g++ 14.1.1 either requires -fcoroutines or -std=gnu++20.

**Comment by wsnyder** [comment]
>Clang (at least apple)

Note we successfully build every commit with clang on Linux, so the comments above do relate not to clang generally, but as to how Apple chooses to (mis-)package clang and cause this breakage.  (Generally Apple breaks so much related to C++ tools it seems it must be intentional.)

Anyhow, CFG_CXXFLAGS_STD_NEWEST isn't currently used because it breaks too much, and relatedly we don't want to generally force C++20.  It is fine however to add -std=gnu++20 to CFG_CXXFLAGS_COROUTINES as something to try (after the current -fcoroutines trials) for experiments for ./configure to check to get coroutines working.

**Comment by gussmith23** [comment]
I realize this is probably not super helpful, but modifying latest Verilator with these brute-force fixes makes things work:
```diff
diff --git a/configure.ac b/configure.ac
index d7621ac01..00d43d4ff 100644
--- a/configure.ac
+++ b/configure.ac
@@ -356,8 +356,8 @@ AC_SUBST(CFG_CXXFLAGS_PROFILE)
 # c++17 is the newest that Verilator is regularly tested to support
 # c++14 is the oldest that Verilator supports
 # gnu is required for Cygwin to compile verilated.h successfully
-#_MY_CXX_CHECK_SET(CFG_CXXFLAGS_STD_NEWEST,-std=gnu++20)
-#_MY_CXX_CHECK_SET(CFG_CXXFLAGS_STD_NEWEST,-std=c++20)
+_MY_CXX_CHECK_SET(CFG_CXXFLAGS_STD_NEWEST,-std=gnu++20)
+_MY_CXX_CHECK_SET(CFG_CXXFLAGS_STD_NEWEST,-std=c++20)
 _MY_CXX_CHECK_SET(CFG_CXXFLAGS_STD_NEWEST,-std=gnu++17)
 _MY_CXX_CHECK_SET(CFG_CXXFLAGS_STD_NEWEST,-std=c++17)
 _MY_CXX_CHECK_SET(CFG_CXXFLAGS_STD_NEWEST,-std=gnu++14)
@@ -376,6 +376,7 @@ AC_SUBST(CFG_CXXFLAGS_WEXTRA)
 _MY_CXX_CHECK_SET(CFG_CXXFLAGS_COROUTINES,-fcoroutines-ts)
 _MY_CXX_CHECK_SET(CFG_CXXFLAGS_COROUTINES,-fcoroutines)
 _MY_CXX_CHECK_SET(CFG_CXXFLAGS_COROUTINES,-fcoroutines-ts -Wno-deprecated-experimental-coroutine)
+_MY_CXX_CHECK_SET(CFG_CXXFLAGS_COROUTINES,-std=gnu++20)
 AC_SUBST(CFG_CXXFLAGS_COROUTINES)
 
 # HAVE_COROUTINES
```

**Comment by wsnyder** [comment]
Does it work with only the second part, changing CXX_FLAGS_COROUTINES?

**Comment by gussmith23** [comment]
No it doesn't -- i meant to delete that line. It seems like that line had no effect.

**Comment by wsnyder** [comment]
Can you make a pull then just with that one 
+_MY_CXX_CHECK_SET(CFG_CXXFLAGS_COROUTINES,-std=gnu++20)

line?

**Comment by gussmith23** [comment]
https://github.com/verilator/verilator/pull/5200

Again, I don't think this is going to work -- I think it's the 
```
+_MY_CXX_CHECK_SET(CFG_CXXFLAGS_STD_NEWEST,-std=gnu++20)
+_MY_CXX_CHECK_SET(CFG_CXXFLAGS_STD_NEWEST,-std=c++20)
```

changes that had the effect!

**Comment by wsnyder** [comment]
Sorry, I misunderstood then.  I can't accept a pull which changes STD_NEWEST.

**Comment by ktbarrett** [comment]
Why not? You are using a feature that isn't guaranteed to be available before C++20 and people are having issues with that lack of guarantee. The problem is pretty clear cut to me.

**Comment by wsnyder** [comment]
Coroutines are *not* required, they are optional, the point of the CFG_CXXFLAGS_COROUTINES is to try some things to see if they can be supported.

**Comment by ktbarrett** [comment]
I guess I'll take a stab at this tonight. I haven't the slightest clue how autoconf works so wish me luck.

**Comment by ktbarrett** [comment]
Well I got busy and haven't worked on it, but I see https://github.com/verilator/verilator/issues/5190. Has anyone confirmed that this fixes the issue?

**Comment by barbedo** [comment]
I just took a look at the current configuration logic, comparing Apple's and upstream Clang:

```autoconf
# Flags for coroutine support for dynamic scheduling
_MY_CXX_CHECK_SET(CFG_CXXFLAGS_COROUTINES,-fcoroutines-ts)
_MY_CXX_CHECK_SET(CFG_CXXFLAGS_COROUTINES,-fcoroutines)
_MY_CXX_CHECK_SET(CFG_CXXFLAGS_COROUTINES,-fcoroutines-ts -Wno-deprecated-experimental-coroutine)
_MY_CXX_CHECK_SET(CFG_CXXFLAGS_COROUTINES,-std=gnu++20)
AC_SUBST(CFG_CXXFLAGS_COROUTINES)

# HAVE_COROUTINES
Check if coroutines are supported at all
AC_MSG_CHECKING([whether coroutines are supported by $CXX with $CFG_CXXFLAGS_COROUTINES])
# ...
```

The issue with current Apple Clang 15.0.0 is that it accepts `-fcoroutines-ts`, but support in libc++ headers is only active for `-std=c++20` or higher. So the first check is successful, even if you cannot use coroutines. `HAVE_COROUTINES` passes because it only includes `<coroutine>`, without using any definition from it.

[Upstream libc++ 17+](https://github.com/llvm/llvm-project/blob/6009708b4367171ccdbf4b5905cb6a803753fe18/libcxx/include/__coroutine/coroutine_traits.h#L19) also needs `-std=c++20` or higher. Latest GNU's libstd++ [only seem to need C++14 and `-fcoroutines`](https://github.com/gcc-mirror/gcc/blob/cd0059a1976303638cea95f216de129334fc04d1/libstdc%2B%2B-v3/include/std/coroutine#L57).

Recently,  [`-fcoroutines-ts` was removed from Clang](https://github.com/llvm/llvm-project/commit/6ed67ccba7e4699e9e42302f2f9b7653444258ba), and `-fcoroutines` throws an "unused" warning, which breaks the current check. So, one effect of #5200 is that it re-enables coroutines for latest Clang:

```
$ clang --version          
clang version 18.1.8
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin

$ CC=clang CXX=clang++ ./configure

# before #5200
checking whether coroutines are supported by clang++... no

# after #5200
checking whether coroutines are supported by clang++... yes
```

**TL;DR**: The situation is a mess and depends on the combination of compilers, their versions, the C++ standard library and its versions.

Getting back to the original issue, one way of solving this (and making the configuration step more robust) is doing the following:
- Couple the coroutine flag check with the coroutine compile check.
  - `HAVE_COROUTINE` is only set in pair with the correct `CFG_CXXFLAGS_COROUTINES`.
- Add a small declaration of a variable of a type from `std::coroutine` in the `HAVE_COROUTINE` check.
  - E.g. `std::coroutine_handle<> h;`
  - The current check only verifies if the header can be included.
- Keep trying the flags until the compile check works.
  - If none of them works, disable coroutines. 

This way, we can have something like this that should hopefully will work for all the combinations:

```
# macOS
# Apple Clang 15
checking whether coroutines are supported by g++ -std=gnu++11... no
checking whether coroutines are supported by g++ -std=gnu++11 with -fcoroutines-ts... no
checking whether coroutines are supported by g++ -std=gnu++11 with -fcoroutines... no
checking whether coroutines are supported by g++ -std=gnu++11 with -fcoroutines-ts -Wno-deprecated-experimental-coroutine... no
checking whether coroutines are supported by g++ -std=gnu++11 with -std=gnu++20... yes

# Ubuntu
# GCC 14
checking whether coroutines are supported by g++-14... no
checking whether coroutines are supported by g++-14 with -fcoroutines-ts... no
checking whether coroutines are supported by g++-14 with -fcoroutines... yes

# For fun, Clang 15 + libc++
checking whether coroutines are supported by clang++-15... no
checking whether coroutines are supported by clang++-15 with -fcoroutines-ts... yes

# Clang 18 + libstd++
checking whether coroutines are supported by clang++-18... no
checking whether coroutines are supported by clang++-18 with -fcoroutines-ts... no
checking whether coroutines are supported by clang++-18 with -fcoroutines... no
checking whether coroutines are supported by clang++-18 with -fcoroutines-ts -Wno-deprecated-experimental-coroutine... no
checking whether coroutines are supported by clang++-18 with -std=gnu++20... yes
```

If this approach is acceptable, I can create a PR.

**Comment by barbedo** [comment]
@ktbarrett, I've just created #5300. Feel free to give it a spin.

**Comment by gussmith23** [comment]
I really appreciate this -- this has been an issue on Mac for the past
year+ for me, and I didn't realize it was a Verilator bug and not just me
misunderstanding the tool. Happy to get a smoother Verilator experience.

**Comment by wsnyder** [comment]
Well done, thanks & merged.

**Comment by ktbarrett** [comment]
I tested locally and the fix works.
